ODLC Text Color Parsing

fallscameron01 opened the pull request:
closes #23 

# Changes

- Rearranged class to increase organization
- Improved code documentation
- Added typing to existing code
- Conformed existing code to new repo standards
- Minor changes to pylint settings
- Added algorithms for parsing standard object text color
- Implementation of function to determine text color
- Implemented mapping of HSV color to color strings in interoperability system

fallscameron01 committed:
Color distance calculation

fallscameron01 committed:
Kmeans implementation

fallscameron01 committed:
Mask of colors

fallscameron01 committed:
Implement distance calculation

fallscameron01 committed:
Rearrange functions

fallscameron01 committed:
Split into multiple functions

fallscameron01 committed:
Fix BGR/RGB issue

fallscameron01 committed:
HSV Color ranges

fallscameron01 committed:
Implement color matching

fallscameron01 committed:
Implement color resolving

fallscameron01 committed:
Switch to member variables, cleanup

fallscameron01 committed:
Fix RGB/BGR issue

fallscameron01 committed:
Typed variables

fallscameron01 committed:
Added more typing

fallscameron01 committed:
Fix Imports

fallscameron01 committed:
Merge branch 'develop' into feature/alphanumeric_color

fallscameron01 committed:
Re-added images to gitignore

fallscameron01 committed:
Conforming to new coding standards

fallscameron01 committed:
Merge branch 'develop' into feature/alphanumeric_color

mrouie commented on vision/Object Detection/standard_obj_text.py:
```diff
@@ -2,76 +2,116 @@
 
         ## Detect Text ##
-        print("Image processing complete.")
         txt_data = pytesseract.image_to_data(
```
Missing type hint for function call return assignment

mrouie commented on vision/Object Detection/standard_obj_text.py:
```diff
@@ -80,79 +120,84 @@ def _detect_text(self, img: np.ndarray, bounds: np.ndarray) -> list:
-                    w = txt_data["width"][i]
-                    h = txt_data["height"][i]
+                    width = txt_data["width"][i]
+                    height = txt_data["height"][i]
```
Make sure you also type hint all of these assignments

mrouie commented on vision/Object Detection/standard_obj_text.py:
```diff
@@ -80,79 +120,84 @@ def _detect_text(self, img: np.ndarray, bounds: np.ndarray) -> list:
+                    height = txt_data["height"][i]
 
                     # Don't continue processing if text is size of full image
                     img_h, img_w = np.shape(processed_img)
```
I know this might seem like an obvious output, but consistency is key, so we gotta type hint this tuple unpacking.
The normal way to do this is one line at a time, but if you want to do this all on one line, you have to use semicolons, (i.e. `a: int; b: int` before the tuple unpacking).

fallscameron01 replied:
Semicolons in Python? What the world has come to.

fallscameron01 replied:
Also should note that the semicolon fix does not work. Black will remove the semicolon and separate it like this:
```
a: int; b: int
```
will become
```
a: int
b: int
```
So, multiline is the only way this can be done.

mrouie replied:
I know right, semicolons sound like the most sacrilegious thing to exist in a language like this. It's the only way you can write multiple statements on one line, and I agree they should be disallowed in nearly all circumstances. Unfortunately, type hinting tuple unpacking can get crazy and add a lot of lines. 
You almost had it correct. Tuple unpacking type annotations always have to be done immediately before the unpacking, so the right way to do it would've been:
```
a: int; b: int
a, b = foo()
```
I thought that black would allow semicolons in this case, but that's my bad. Black actually completely doesn't allow them (understandably) for the reason that they are not pythonic. There's also no way for me to add an exception for semicolons. So, you're gonna have to do it the main way I had it in my slides, which is the way [PEP 526](https://www.python.org/dev/peps/pep-0526/#global-and-local-variable-annotations) says to do it, which unfortunately uses an extra line for each element of the tuple
```
a: int
b: int
a, b = foo()
```

mrouie commented on vision/Object Detection/standard_obj_text.py:
```diff
@@ -80,79 +120,84 @@ def _detect_text(self, img: np.ndarray, bounds: np.ndarray) -> list:
-                    if not (x == 0 and y == 0 and w == img_w and h == img_h):
-                        bounds = [(x, y), (x + w, y), (x + w, y + h), (x, y + h)]
+                    if not (x == 0 and y == 0 and width == img_w and height == img_h):
+                        t_bounds = [
```
Also type hint this list

mrouie commented on vision/Object Detection/standard_obj_text.py:
```diff
@@ -80,79 +120,84 @@ def _detect_text(self, img: np.ndarray, bounds: np.ndarray) -> list:
+        laplace_img = cv2.Laplacian(dilated, ddepth=cv2.CV_8U, ksize=5)
+
+        # binarize image
+        binarized = np.where(laplace_img > 50, np.uint8(255), np.uint8(0))
```
pls type hint all of these :)

mrouie commented on vision/Object Detection/standard_obj_text.py:
```diff
@@ -182,53 +227,229 @@ def _slice_rotate_img(self, img: np.ndarray, bounds: np.ndarray) -> np.ndarray:
-        gray = cv2.cvtColor(img, cv2.COLOR_RGB2GRAY)
+        # Slice rotated image around bounds of text ##
+        x_vals = [coord[0] for coord in char_bounds]
+        y_vals = [coord[1] for coord in char_bounds]
```
These list comps need type hints as well

mrouie commented on vision/Object Detection/standard_obj_text.py:
```diff
@@ -182,53 +227,229 @@ def _slice_rotate_img(self, img: np.ndarray, bounds: np.ndarray) -> np.ndarray:
+
+        Returns
+        -------
+        None
```
Returns documentation isn't needed for a function that never returns anything.
In this case, a single summary docstring suffices.

mrouie commented on vision/Object Detection/standard_obj_text.py:
```diff
@@ -182,53 +227,229 @@ def _slice_rotate_img(self, img: np.ndarray, bounds: np.ndarray) -> np.ndarray:
+        vectorized = np.hstack((vectorized, idxs))
+
+        # Run Kmeans with K=2
+        term_crit = (
```
type hint this tuple as well :)

mrouie commented on vision/Object Detection/standard_obj_text.py:
```diff
@@ -182,53 +227,229 @@ def _slice_rotate_img(self, img: np.ndarray, bounds: np.ndarray) -> np.ndarray:
+            1.0,
+        )
+        k_val: int = 2
+        _, label, center = cv2.kmeans(
```
also type hint this tuple unpacking pls

mrouie commented on vision/Object Detection/standard_obj_text.py:
```diff
@@ -182,53 +227,229 @@ def _slice_rotate_img(self, img: np.ndarray, bounds: np.ndarray) -> np.ndarray:
+            the color of the text
+        """
+        ## Find the two colors in the image ##
+        img_colors: np.ndarray = np.unique(
```
use numpy.typing with structural subtyping for this

mrouie commented on vision/Object Detection/standard_obj_text.py:
```diff
@@ -182,53 +227,229 @@ def _slice_rotate_img(self, img: np.ndarray, bounds: np.ndarray) -> np.ndarray:
+
+        # average distance for each color
+        dist_1 = cv2.mean(distance_mat, color_1_mat)[0]
+        dist_2 = cv2.mean(distance_mat, color_2_mat)[0]
```
type these also

mrouie commented on vision/Object Detection/standard_obj_text.py:
```diff
@@ -182,53 +227,229 @@ def _slice_rotate_img(self, img: np.ndarray, bounds: np.ndarray) -> np.ndarray:
+                        np.sum(np.abs(hsv_color_val - mid2)),
+                    )
+                else:  # any color except red
+                    mid = np.mean(POSSIBLE_COLORS[col])  # midpoint of range
```
all of best_dist, dist, mid1, mid2, and mid need type hints on their first/outermost initial assignments

mrouie commented on vision/Object Detection/standard_obj_text.py:
```diff
@@ -182,53 +227,229 @@ def _slice_rotate_img(self, img: np.ndarray, bounds: np.ndarray) -> np.ndarray:
+        ## TEMP: Implemmentation out of scope of current issue
+        self.orientation = "N"
+
+        return self.orientation
```
Unimplemented functions should should raise the NotImplementedError only.
If you want, you can comment out the code and add the error. You also don't need to add a Raises section to the docstring if its only for a NotImplementedError. Also, pylint & mypy should recognize that as an unfinished function, so you should be fine having a return type hint with no return value.

mrouie commented on vision/Object Detection/standard_obj_text.py:
```diff
@@ -249,13 +470,13 @@ def _preprocess_img(self, img: np.ndarray) -> np.ndarray:
         raise RuntimeError("No file specified.")
     file_name = args.file_name
```
new variable in scope, needs type hint

mrouie commented on vision/Object Detection/standard_obj_text.py:
```diff
@@ -182,53 +227,229 @@ def _slice_rotate_img(self, img: np.ndarray, bounds: np.ndarray) -> np.ndarray:
-    """
     import argparse
 
     parser = argparse.ArgumentParser(
```
Please type this. Yes we are typing literally everything lol
However, you do have to type the `args` Namespace that comes later

mrouie commented on vision/Object Detection/standard_obj_text.py:
```diff
@@ -249,13 +470,13 @@ def _preprocess_img(self, img: np.ndarray) -> np.ndarray:
 
     detector = TextCharacteristics()
-    detected_chars = detector.get_text_characteristics(img, bounds)
+    detected_chars = detector.get_text_characteristics(test_img, test_bounds)
```
User-defined class instantiations should be typed as well as this method call return

mrouie commented on vision/Object Detection/standard_obj_text.py:
```diff
@@ -2,76 +2,116 @@
+        img : np.NDArray[np.uint8]
             image to find characteristics of text within
-        bounds : np.ndarray
+        bounds : List[List[int, int]]
```
Make sure you describe the expected format of bounds list at the first instance that the concept is referenced in code. Currently, it's described most thoroughly at line 196.

I also noticed you have a bit of an inconsistency with your definition of "bounds" as in this part of your code and in main, you refer to bounds as a `List[List[int]]`, but in some other parts like line 75, you refer to bounds as a `List[Tuple[int, int]]`. It's safe to assume that the user of these methods will be passing at the very least a `List[Tuple[int, int]]` since it makes more sense to use an immutable type for coordinates..

mrouie commented on vision/Object Detection/standard_obj_text.py:
```diff
@@ -80,79 +120,84 @@ def _detect_text(self, img: np.ndarray, bounds: np.ndarray) -> list:
+        None
         """
         ## Slice image around bounds and find center point ##
         x_vals = [coord[0] for coord in bounds]
```
everything in this function needs type hinting as well

mrouie commented on vision/Object Detection/standard_obj_text.py:
```diff
@@ -2,76 +2,116 @@
         -------
-        tuple
-            characteristics of the text in the form (character, orientation, color)
+        (character, orientation, color) : Tuple[str, str, str]
```
This is not just a tuple of strings, but a tuple of Optional strings, so you should label it as such (either with `Optional[str]` or `str | None` for each str in the tuple, describing what the meaning of a None type output means for each value of the tuple

mrouie commented on vision/Object Detection/standard_obj_text.py:
```diff
@@ -2,76 +2,116 @@
         -------
-        list
-            list containing detected characters and their bounds, Format of characters is ([bounds], 'character').
+        found_characters : List[Tuple[str, List[Tuple[int, int]]]]
```
You might want to extrapolate this structural subtyping at least one level further. 
If you want, as a way to clean things up, you can utilize `TypeAlias` or `NewType` for things like `List[Tuple[int, int]]` since it is used quite a few times in this module, but you don't have to do this.
The goal here is to make it clear how a collection of collections of collections actually maps to the logical output format. (i.e. where is a found character and bounds in this complex return type, and what do bounds mean/look like.)

mrouie commented on vision/Object Detection/standard_obj_text.py:
```diff
@@ -80,79 +120,84 @@ def _detect_text(self, img: np.ndarray, bounds: np.ndarray) -> list:
+        bounds : List[List[int]]
+            array of bound coordinates (4 x-y coordinates; tl-tr-br-bl)
 
         Returns
```
Returns section not needed if there are no returns.

mrouie commented on vision/Object Detection/standard_obj_text.py:
```diff
@@ -182,53 +227,229 @@ def _slice_rotate_img(self, img: np.ndarray, bounds: np.ndarray) -> np.ndarray:
-        np.ndarray
-            the image after preprocessing
+        color : str
+            the color of the text
```
I would reference in some way all the possible colors that can be output. If you want you can use the discrete output types as a set using `retstring : {"val1", "val2", "val3", "val4"}` instead of `retstring : str` or you can simply just reference your possible colors dictionary that you made.

Also, you need to mention here that color could be None (with `Optional[str]` or `str | None` and what a None type color logically means

mrouie reviewed (changes requested):
Overall, your code structure is well-written and user-friendly.
However, we do need to be type annotating every initial instance of a variable or constant in any new scope, and some of your documentation could use a little bit more detail.
Other than that, as long as the code is functionally sound, it'll be OK to merge.

fallscameron01 committed:
Fixed typing issues

mrouie commented on .pylintrc:
```diff
@@ -523,7 +523,7 @@ max-returns=6
 
 # Minimum number of public methods for a class (see R0903).
-min-public-methods=2
+min-public-methods=1
```
Ok so I did some further review of your program and proper programming practices, and I made the conclusion that we should keep this at 2 minimum public methods for a class.

mrouie reviewed (changes requested):
This marks the more generalized review of why we should keep the minimum public methods to 2. The reasoning is that, in 95% of cases (unless you're in a language like Java), classes with only one public method are just functions in disguise. There are some cases where 0 public methods makes sense (for example in the cases of named tuples, dataclasses, and attr classes).

For the case of your text characteristics module, I found it a bit confusing to need to create an empty-constructed class, then call a function where you pass all the data and receive all the resultant data. The implication here is that the class exists, but it seems like the user isn't really intended to use any of its instance variables or instance methods except for the one (`get_text_characteristics`) method. It's basically not very user friendly from a functional programming perspective.

If you want to retain the class structure, then it needs to be justified with usable attributes/methods.
One way you could do this is by using a getter `@property` for each of the characteristics and implementing a constructor that gets passed the relevant information so that the methods can operate on that data.

fallscameron01 committed:
Initial BoundingBox object

fallscameron01 committed:
Added BoundingBox helper functions

fallscameron01 committed:
Changed vertices to property

fallscameron01 commented:
closes #50

fallscameron01 committed:
Changed obj_type and attributes to properties

fallscameron01 committed:
Switched types to BoundingBoxes

fallscameron01 committed:
Change back to 2 min public methods

fallscameron01 committed:
Restructure class

fallscameron01 committed:
Change ObjectType to strings

fallscameron01 committed:
Change to private variables

fallscameron01 committed:
Changed rotated_img to property

fallscameron01 commented:
**New Changes:**

- Fixed requested typing issues
- Created BoundingBox class for use in vision algorithms
- Changed TextCharacteristics functions to use new BoundingBox object
- Change pylint back to 2 minimum public methods
- Made structural changes to TextCharacteristics class to conform to 2 minimum public methods
- Changed member variables of TextCharacteristics class to private
- Added init_hook to pylint to fix local import issues
- Removed disallow generics rule from mypy

fallscameron01 committed:
Change attribute type

fallscameron01 committed:
Disallow any generics

mrouie reviewed (approved):
Good stuff.
I like the more enriched bounding box class with the new methods you added to add abstract away some of the bounding-box-specific operations you were doing. I'm not really sure having getters/setters for the rotated image is actually necessary, but if you think it could be used somewhere, then it's fine.
